docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+    - update flags for the dynamic pages
+
+## munmap 
+Calling munmap on an enclave page (dynamic allocated or not) has exactly the same effect of calling munmap on a regular RAM page. No sgx specific interface is needed. No behavior changes to current kernel space implementation.
+
+### Remarks
+
+- Enclave memory mapings are shared (MAP_SHARED). The mappings in shared processes are kept alive and independently until the process exits
+    - munmap and closing file descriptors are not required for user space. A dead process automatically releases all mappings and file descriptors. 
+- Upon all enclave mappings are removed and file handles to the enclave are closed, either by explicit munmap/fclose syscalls or when all hosting apps exited:
+    - The kernel may mark its remaining pages are reclaimable and issue EREMOVE on them any time the kernel deems appropriate.
+
+## mprotect IOCTL
+This IOCTL emulates the mprotect syscall with SGX specific extensions. In future kernel implementation, it could be mprotect or pkey_mprotect syscall with sgx extensions for the "prot" parameter.
+
+```
+#define SGX_IOC_ENCLAVE_MPROTECT _IOW(SGX_MAGIC, 0x06, struct sgx_enclave_mprotect)
+/**
+ * struct sgx_enclave_mprotect - parameter structure for the
+ *                                %SGX_IOC_ENCLAVE_MPROTECT ioctl
+ * @addr:	address of the memory to change protections
+ * @length:	length of the area.  This must be a multiple of the page size.
+ * @prot:	this must be or'ed of following:
+             PROT_READ
+             PROT_WRITE
+             PROT_EXEC
+             PROT_TRIM (new): change the page type to PT_TRIM, implies RW. User space should immediately EACCEPT, and then optionally call mprotect with PROT_EACCEPT.
+             PROT_TCS (new): change the page type to PT_TCS
+             PROT_EACCEPT (new): notify kernel EACCEPT is done for PT_TRIM type changes. This can be optional. Kernel will try EREMOVE on PT_TRIM pages in best effort on its own if user space does not notify EACCEPT.
+ */
+struct sgx_enclave_mprotect {
+	__u64	addr;
+	__u64	length;
+	__u64	prot;
+};
+```
+
+### Remarks
+
+Kernel should ensure that SGX instructions can succeed or catch and handle any fault.
+  - The kernel may maintain EPCM information on each page which includes access permission RWX, page types of PT_REG, PT_TRIM, PT_TCS.
+  - The kernel catches fault on EMODPR, EMODT and converts to error code returned to user space.
+
+The enclave run-time (or trusted run-time) may implement a parallel memory management structure which would provide information to the enclave on the enclave memory mappings.  The run-time can have a trusted API analogous to mmap which makes a call out of the enclave to issue the mmap  and then either perform EACCEPT on the pages and update the internal memory structures or configure the enclave to perform the EACCEPT when a #PF is delivered to the enclave.  With the outlined kernel interface, either implementation is possible.
+
```

Comment:
Relying on page faults and error handling to add new pages to the enclave is ill-suited in some scenarios. The `mmap` approach can only provide a hint where a new page should be added, not enforce a specific location. A `syscall/ioctl` to issue an `eaug` instruction for a specific page would make much more sense.
Consider as an example an enclave written mostly in a memory safe language, with a small fraction of C/assembly. When such an enclave issues a page fault due to a page not being present, something bad must have happened, and the enclave should stop execution immediately. Letting it add new pages as default behavior may make successful exploitation of memory safety vulnerabilities much more likely.

Reply:
First we are not proposing adding pages in error handling path. It is usual practice for kernel to allocate pages upon #PF. Kernel also can decide immediately to allocate EPC on mmap if it sees fit based on hint from user space and overall resource usage of the system. 

Second I don't see how direct EAUG would be better suited to mitigate the security problem you stated. You can use mmap/mprotect to specify  regions  allowed to add pages with specific access permissions allowed, which should effectively prevent ill-intended faults. Note, without EACCEPT, even if  a page is EAUG'd, app can't use the page. So user space can always control which page to be EACCEPT'd, further prevent the scenario you described.

Reply:
@raoulstrackx I would also recommend to check the original "Intel® Software Guard Extensions (Intel® SGX) Software
Support for Dynamic Memory Allocation inside an Enclave" paper. These flows are described very well there.